cargo <1.60 failed to select a version of a dependency due to either `?` (weak) or `dep:` (namespaced) features

### Problem

I did a `cargo update` and my CI broke when verifying MSRV. 
```console
error: failed to select a version for the requirement `libgit2-sys = "^0.13.3"`
candidate versions found which didn't match: 0.13.2+1.4.2, 0.13.1+1.4.2, 0.13.0+1.4.1, ...
location searched: crates.io index
required by package `git2 v0.14.3`
    ... which satisfies dependency `git2 = "^0.14"` (locked to 0.14.3) of package `repor v0.1.0 (/home/epage/src/personal/repro)`
```

- cargo 1.60 works
- cargo 1.59 doesn't work

The error being from the resolver and being cargo-version specific is why I am posting here rather than on the git2 repo.

### Steps

```console
$ git clone git@github.com:crate-ci/committed.git
cd committed
cargo +1.59.0 check
cargo +1.60.0 update
cargo +1.59.0 check
```

### Possible Solution(s)

Track parse bad candidates when resolving and report them to the user if they are relevant

### Notes

Old cargo cannot see crates in the registry that use [weak or namespaced features](https://rust-lang.github.io/rfcs/3143-cargo-weak-namespaced-features.html) (or similar new index data).

### Version

```text
cargo 1.60.0 (d1fd9fe 2022-03-01)
release: 1.60.0
commit-hash: d1fd9fe2c40a1a56af9132b5c92ab963ac7ae422
commit-date: 2022-03-01
host: x86_64-unknown-linux-gnu
libgit2: 1.3.0 (sys:0.13.23 vendored)
libcurl: 7.80.0-DEV (sys:0.4.51+curl-7.80.0 vendored ssl:OpenSSL/1.1.1m)
os: Linux Mint 20.3 (una) [64-bit]
```

> Considering the usability issue of this, re-opening for us providing better errors.

Below examples are a bit verbose, but may provide some value? 🤷‍♂️ (_I'm mostly documenting the experience_)

---

The `hashbrown 0.14.2` example differs slightly from your error output, in that:
- It informs the user that it could not satisfy `0.14.2` semver, but found `0.14.2` version of the crate.
- Actual failure was due to not being able to satisfy `-Z msrv-policy`.

I don't really use such old toolchain releases, but was contributing to a crate which did where I encountered this resolver confusion (_later discovering `-Z msrv-policy`_).
- Initially I thought it was due to cargo resolving compatible crates like `hashbrown` by the `rust-version` and rolling back parent dependencies until satisfied (_similar to `-Z msrv-policy` but not quite the same_).
- Later it was clarified on Discord to me that this was due to the `1.60.0` changes to `Cargo.toml` being adopted, making those crates invisible to older toolchains.

I'm not sure if this sort of incompatibility has also been introduced again in newer toolchain releases since too? (_It was definitely unexpected_ 😅 )

---

### Example - Resolving `ordered-multimap = "0.4"` with Rust `1.56.1`

#### Resolving with `rust-version` via `-Z msrv-policy` / `cargo-upgrade`

If `ahash` had already adopted `rust-version`, this could have resolved correctly (_**EDIT:** it will declare [`rust-version = "1.60.0"` for future releases](https://github.com/tkaitchuck/aHash/pull/181)_):

```console
# `Cargo.toml` with single dep of `ordered-multimap = "0.4"` and `rust-version = "1.56.1"`:
# No failure encountered implies it was successful meeting the MSRV requirement (or at least on best effort)
$ cargo +nightly update -Z msrv-policy


# hashbrown respecting rust-version, resolves ahash semver (assuming compatibility with hashbrown rust-version)
# - 0.12.3 has rust-version 1.56.0, and `ahash = "0.7"` semver
# - 0.11.2 is similar, but published before `ahash 0.7.6` was released (which implicitly raised MSRV to 1.60)
$ cargo tree
app v0.1.0 (/app)
└── ordered-multimap v0.4.3
    ├── dlv-list v0.3.0
    └── hashbrown v0.12.3
        └── ahash v0.7.7
            ├── getrandom v0.2.10
            │   ├── cfg-if v1.0.0
            │   └── libc v0.2.149
            └── once_cell v1.17.2
            [build-dependencies]
            └── version_check v0.9.4


# Fails because `ahash >= 0.7.6` uses `Cargo.toml` syntax that can't be parsed by Rust <1.60.0:
$ cargo +1.56.1 check
    Updating crates.io index
error: failed to select a version for the requirement `ahash = "=0.7.7"`
candidate versions found which didn't match: 0.4.8, 0.4.1, 0.4.0, ...
location searched: crates.io index
required by package `hashbrown v0.12.3`
    ... which satisfies dependency `hashbrown = "=0.12.3"` of package `ordered-multimap v0.4.3`
    ... which satisfies dependency `ordered-multimap = "=0.4.3"` of package `app v0.1.0 (/app)`


# Or with cargo-edit installed, and a `ordered-multimap = "0.4.0" semver:
# cargo-upgrade uses same `-Z msrv-policy` approach AFAIK (but applies changes to `Cargo.toml`, raising min version)
$ cargo +1.56.1 upgrade
    Updating 'https://github.com/rust-lang/crates.io-index' index
    Checking app's dependencies
name             old req compatible latest new req note
====             ======= ========== ====== ======= ====
ordered-multimap 0.4.0   0.4.3      0.7.1  0.4.3   incompatible
   Upgrading recursive dependencies
note: Re-run with `--incompatible` to upgrade incompatible version requirements
```

#### Resolving with `cargo +1.56.1 check` (_without a pre-existing `Cargo.lock`_)

This time `rust-version` doesn't affect resolving, but the toolchain being below `1.60.0` notably affects the resolver (silently):

```console
$ cargo +1.56.1 check
    Updating crates.io index
    Finished dev [unoptimized + debuginfo] target(s) in 1.65s

$ cargo tree
app v0.1.0 (/app)
└── ordered-multimap v0.4.0
    ├── dlv-list v0.2.3
    │   └── rand v0.8.5
    │       ├── libc v0.2.149
    │       ├── rand_chacha v0.3.1
    │       │   ├── ppv-lite86 v0.2.17
    │       │   └── rand_core v0.6.4
    │       │       └── getrandom v0.2.10
    │       │           ├── cfg-if v1.0.0
    │       │           └── libc v0.2.149
    │       └── rand_core v0.6.4 (*)
    └── hashbrown v0.9.1
        └── ahash v0.4.8
```

- Resolves to `ordered-multimap 0.4.0` instead, with `hashbrown 0.9.1` satisfying the semver `ahash ="0.4.4"` which resolves to the `ahash 0.4.8` release.
  
  <details>
  <summary>Ignore these points regarding `once_cell` (unrelated)</summary>
  
  (_**EDIT:** [With `default-features = false`](https://github.com/rust-lang/hashbrown/blame/adf06c2e7f3371ffdbff04388d3894978fdf2bf4/Cargo.toml#L16), so ignore the `once_cell` sub-points that follow as [that depends upon the "compile-time-rng" feature installing `const-random`](https://github.com/tkaitchuck/aHash/blob/5c08915b4314e3bf6e37600b2af93f5fcdde7d31/Cargo.toml#L22)_)  
  
  - `ahash 0.4.8` technically resolves to a MSRV of `1.60.0` due to an implicit dependency semver resolving to `once_cell 1.18.0` (_detected via `cargo-msrv`_) which [adopted the `dep:` feature](https://github.com/matklad/once_cell/blob/65568632484c5e108688e916a1e04d2806aedb26/Cargo.toml#L51). However since these Rust `1.60.0` changes prevent resolving that on earlier toolchains, `1.17.2` and below are compatible to resolve for `1.56.1` instead (_without needing `-Z msrv-policy`_).
  - The actual MSRV for `ahash 0.4.8` is as low as Rust `1.56.0` (via `-Z minimal-versions --package once_cell`) - since the [min semver bound for `once_cell`](https://github.com/tkaitchuck/constrandom/blob/e9f560ba14e09ff9db9caca3f2dfa3ff52cc96de/macro/Cargo.toml#L18) is set to `1.15` (_which [introduced `edition = "2021"`](https://github.com/matklad/once_cell/blob/97edd07e0ac05fb8e4b994e9a53ba187d8fa17fc/Cargo.toml#L6) + adopted `rust-version` from that release onwards_).
  
  </details>
  
- The semver range `ordered-multimap = "0.4"` could not resolve to newer releases of `0.4.x`, due to those releases bumping minor version of `hashbrown` (_which also bumped `ahash` minors that relied on the since yanked `ahash` releases, or releases with incompatible `dep:` syntax in `Cargo.toml` that requires Rust `>=1.60.0`_).
- `ahash` could publish a new release with an appropriate `rust-version` for a better error message, but that would be too late now as it'd resolve to `ahash 0.7.7` as the release prior to declaring `rust-version` (_if no release with `rust-version` declared is compatible_).

---

### Example - `hashbrown 0.14.2` leveraging `rust-version`

Whereas if our `Cargo.toml` (_with `rust-version = "1.56.1"`_) instead uses `hashbrown = "0.14.2"` (_latest version, which [has `rust-version = "1.63.0"`](https://github.com/rust-lang/hashbrown/blob/7cd193accc21c51c3d5e2b2b5af4bf558ab859d3/Cargo.toml#L13)_), the failure is caught when using `-Z msrv-policy` this time:

```console
# Not a helpful failure message, `cargo check` is better at communicating when the MSRV of a crate is too high:
$ cargo +nightly update -Z msrv-policy
    Updating crates.io index
error: failed to select a version for the requirement `hashbrown = "^0.14"`
candidate versions found which didn't match: 0.14.2, 0.14.1, 0.14.0, ...
location searched: crates.io index
required by package `app v0.1.0 (/app)`
perhaps a crate was updated and forgotten to be re-vendored?


$ cargo tree
app v0.1.0 (/app)
└── hashbrown v0.14.2
    ├── ahash v0.8.6
    │   ├── cfg-if v1.0.0
    │   ├── once_cell v1.18.0
    │   └── zerocopy v0.7.20
    │   [build-dependencies]
    │   └── version_check v0.9.4
    └── allocator-api2 v0.2.16


# Better failure message (1.60.0 used to avoid `Cargo.toml` syntax incompatibility):
$ cargo +1.60.0 check
error: package `hashbrown v0.14.2` cannot be built because it requires rustc 1.63.0 or newer, while the currently active rustc version is 1.60.0

# But not if the chosen toolchain lacks the newer `Cargo.toml` syntax support (version doesn't appear to exist):
$ cargo +1.56.1 check
error: failed to select a version for the requirement `hashbrown = "=0.14.2"`
candidate versions found which didn't match: 0.13.2, 0.13.1, 0.12.3, ...
location searched: crates.io index
required by package `app v0.1.0 (/app)`
```

`-Z msrv-policy` not being stable, the error message there is perhaps less of a concern?
- ❌ Failure was due to semver not covering a release compatible with the requested `rust-version`.
- ❌ Error message confusingly lacks the context, stating it found `0.14.2` but that it didn't match the `0.14.2` semver requirement.

---

#### Example - `rust-version` is only reliable when constraints to support it are known

<details>
<summary>Collapsed as slightly off-topic</summary>

This is only referencing `hashbrown` as it was part of the prior `hashbrown` troubleshooting with understanding how to leverage `rust-version`. I'm sure the issue described below will be applicable to other crates and their dependency chains until a common `rust-version` baseline is adopted more broadly in the ecosystem (_perhaps with `edition = "2024"`_).

Unfortunately, as `rust-version` appears to be more of a guide for the MSRV, it does not necessarily mean it'll be successful with the default features (_or remain successful in future due to implicit dependency updates within the semver bound_):
- `hashbrown 0.14.x` started with `rust-version = "1.64.0`, then [`hashbrown 0.14.2` relaxed this to Rust `1.63.0`](https://github.com/rust-lang/hashbrown/pull/457#issue-1864074133) for the benefit of Debian 12 Bookworm (_current Debian stable with Rust `1.63`_) with the caveat that [requires version pinning `allocator-api2 0.2.9` in `Cargo.lock`](https://github.com/rust-lang/hashbrown/commit/35e6124af159ce483991fab1b79c14d6238b452d#diff-da0a9b3bf0317fc5ba5c008f9a789601ae025f6e3e7451a0f9193e2ce22b38e7R20) to support.
- That's not clearly documented beyond the commit message referenced by `git blame`, or the [vague changelog item of relaxing the MSRV](https://github.com/rust-lang/hashbrown/commit/f08765f6ce20d45dd336aeb3e7ed3e6e98bbf9dc#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR20).

```console
# Build with toolchain specified in hashbrown `rust-version`:
$ cargo +1.63.0 check
...
error[E0658]: use of unstable library feature 'core_c_str'
...
For more information about this error, try `rustc --explain E0658`.
error: could not compile `allocator-api2` due to 6 previous errors
warning: build failed, waiting for other jobs to finish...
```

Without tracking down the latest compatible version, the minimum version specified by `hashbrown` can at least be relied upon:

```console
# Cargo.lock update attempt based on compile error:
$ cargo +nightly update --package allocator-api2 -Z minimal-versions
    Updating crates.io index
 Downgrading allocator-api2 v0.2.16 -> v0.2.9

# Success
$ cargo +1.63.0 check
    Checking hashbrown v0.14.2
    Checking app v0.1.0 (/app)
    Finished dev [unoptimized + debuginfo] target(s) in 0.36s
```

Or opt-out of the [default `allocator-api2` dependency / feature](https://github.com/rust-lang/hashbrown/blame/778e235eccc233f2d8b906ae8e4c3d717c44fcce/Cargo.toml#L50) in `Cargo.toml` if viable (_usually more inconvenient:_ https://github.com/rust-lang/cargo/issues/3126 ):

```toml
# Opt-out of default feature "allocator-api2":
hashbrown = { version = "0.14.2", default-features = false, features = ["ahash", "inline-more"] }
```

That probably would have been preventable with `-Z msrv-policy` too - if `allocator-api2 0.2.10` had declared `rust-version = "1.63.0"` compatibility (_prior to the `0.2.11` release, which should also have set `rust-version = "1.64.0"`_).

</details>

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -33,7 +33,7 @@ cargo-credential-macos-keychain = { version = "0.4.7", path = "credential/cargo-
 cargo-credential-wincred = { version = "0.4.7", path = "credential/cargo-credential-wincred" }
 cargo-platform = { path = "crates/cargo-platform", version = "0.2.0" }
 cargo-test-macro = { version = "0.4.0", path = "crates/cargo-test-macro" }
-cargo-test-support = { version = "0.6.0", path = "crates/cargo-test-support" }
+cargo-test-support = { version = "0.7.0", path = "crates/cargo-test-support" }
 cargo-util = { version = "0.2.14", path = "crates/cargo-util" }
 cargo-util-schemas = { version = "0.7.0", path = "crates/cargo-util-schemas" }
 cargo_metadata = "0.19.0"

--- a/crates/cargo-test-support/Cargo.toml
+++ b/crates/cargo-test-support/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "cargo-test-support"
-version = "0.6.1"
+version = "0.7.0"
 edition.workspace = true
 rust-version = "1.83"  # MSRV:1
 license.workspace = true

--- a/crates/cargo-test-support/src/registry.rs
+++ b/crates/cargo-test-support/src/registry.rs
@@ -570,7 +570,7 @@ pub struct Package {
     features: FeatureMap,
     local: bool,
     alternative: bool,
-    invalid_json: bool,
+    invalid_index_line: bool,
     edition: Option<String>,
     resolver: Option<String>,
     proc_macro: bool,
@@ -1251,7 +1251,7 @@ impl Package {
             features: BTreeMap::new(),
             local: false,
             alternative: false,
-            invalid_json: false,
+            invalid_index_line: false,
             edition: None,
             resolver: None,
             proc_macro: false,
@@ -1422,8 +1422,8 @@ impl Package {
 
     /// Causes the JSON line emitted in the index to be invalid, presumably
     /// causing Cargo to skip over this version.
-    pub fn invalid_json(&mut self, invalid: bool) -> &mut Package {
-        self.invalid_json = invalid;
+    pub fn invalid_index_line(&mut self, invalid: bool) -> &mut Package {
+        self.invalid_index_line = invalid;
         self
     }
 
@@ -1496,7 +1496,7 @@ impl Package {
             let c = t!(fs::read(&self.archive_dst()));
             cksum(&c)
         };
-        let name = if self.invalid_json {
+        let name = if self.invalid_index_line {
             serde_json::json!(1)
         } else {
             serde_json::json!(self.name)

--- a/crates/cargo-test-support/src/registry.rs
+++ b/crates/cargo-test-support/src/registry.rs
@@ -571,6 +571,7 @@ pub struct Package {
     local: bool,
     alternative: bool,
     invalid_index_line: bool,
+    index_line: Option<String>,
     edition: Option<String>,
     resolver: Option<String>,
     proc_macro: bool,
@@ -1252,6 +1253,7 @@ impl Package {
             local: false,
             alternative: false,
             invalid_index_line: false,
+            index_line: None,
             edition: None,
             resolver: None,
             proc_macro: false,
@@ -1427,6 +1429,14 @@ impl Package {
         self
     }
 
+    /// Override the auto-generated index line
+    ///
+    /// This can give more control over error cases than [`Package::invalid_index_line`]
+    pub fn index_line(&mut self, line: &str) -> &mut Package {
+        self.index_line = Some(line.to_owned());
+        self
+    }
+
     pub fn links(&mut self, links: &str) -> &mut Package {
         self.links = Some(links.to_string());
         self
@@ -1496,22 +1506,26 @@ impl Package {
             let c = t!(fs::read(&self.archive_dst()));
             cksum(&c)
         };
-        let name = if self.invalid_index_line {
-            serde_json::json!(1)
+        let line = if let Some(line) = self.index_line.clone() {
+            line
         } else {
-            serde_json::json!(self.name)
+            let name = if self.invalid_index_line {
+                serde_json::json!(1)
+            } else {
+                serde_json::json!(self.name)
+            };
+            create_index_line(
+                name,
+                &self.vers,
+                deps,
+                &cksum,
+                self.features.clone(),
+                self.yanked,
+                self.links.clone(),
+                self.rust_version.as_deref(),
+                self.v,
+            )
         };
-        let line = create_index_line(
-            name,
-            &self.vers,
-            deps,
-            &cksum,
-            self.features.clone(),
-            self.yanked,
-            self.links.clone(),
-            self.rust_version.as_deref(),
-            self.v,
-        );
 
         let registry_path = if self.alternative {
             alt_registry_path()

--- a/src/cargo/core/resolver/errors.rs
+++ b/src/cargo/core/resolver/errors.rs
@@ -271,6 +271,13 @@ pub(super) fn activation_error(
                         );
                     }
                 }
+                IndexSummary::Invalid(summary) => {
+                    let _ = writeln!(
+                        &mut msg,
+                        "  version {}'s index entry is invalid",
+                        summary.version()
+                    );
+                }
             }
         }
     } else if let Some(candidates) = alt_versions(registry, dep) {

--- a/src/cargo/sources/registry/index/mod.rs
+++ b/src/cargo/sources/registry/index/mod.rs
@@ -135,6 +135,8 @@ pub enum IndexSummary {
     Offline(Summary),
     /// From a newer schema version and is likely incomplete or inaccurate
     Unsupported(Summary, u32),
+    /// An error was encountered despite being a supported schema version
+    Invalid(Summary),
 }
 
 impl IndexSummary {
@@ -144,7 +146,8 @@ impl IndexSummary {
             IndexSummary::Candidate(sum)
             | IndexSummary::Yanked(sum)
             | IndexSummary::Offline(sum)
-            | IndexSummary::Unsupported(sum, _) => sum,
+            | IndexSummary::Unsupported(sum, _)
+            | IndexSummary::Invalid(sum) => sum,
         }
     }
 
@@ -154,7 +157,8 @@ impl IndexSummary {
             IndexSummary::Candidate(sum)
             | IndexSummary::Yanked(sum)
             | IndexSummary::Offline(sum)
-            | IndexSummary::Unsupported(sum, _) => sum,
+            | IndexSummary::Unsupported(sum, _)
+            | IndexSummary::Invalid(sum) => sum,
         }
     }
 
@@ -164,6 +168,7 @@ impl IndexSummary {
             IndexSummary::Yanked(s) => IndexSummary::Yanked(f(s)),
             IndexSummary::Offline(s) => IndexSummary::Offline(f(s)),
             IndexSummary::Unsupported(s, v) => IndexSummary::Unsupported(f(s), v.clone()),
+            IndexSummary::Invalid(s) => IndexSummary::Invalid(f(s)),
         }
     }
 
@@ -280,6 +285,22 @@ impl IndexPackage<'_> {
         summary.set_checksum(self.cksum.clone());
         Ok(summary)
     }
+}
+
+#[derive(Deserialize, Serialize)]
+struct IndexPackageMinimum {
+    name: InternedString,
+    vers: Version,
+}
+
+#[derive(Deserialize, Serialize, Default)]
+struct IndexPackageRustVersion {
+    rust_version: Option<RustVersion>,
+}
+
+#[derive(Deserialize, Serialize, Default)]
+struct IndexPackageV {
+    v: Option<u32>,
 }
 
 /// A dependency as encoded in the [`IndexPackage`] index JSON.
@@ -729,10 +750,45 @@ impl IndexSummary {
         // between different versions that understand the index differently.
         // Make sure to consider the INDEX_V_MAX and CURRENT_CACHE_VERSION
         // values carefully when making changes here.
-        let index: IndexPackage<'_> = serde_json::from_slice(line)?;
+        let index_summary = (|| {
+            let index = serde_json::from_slice::<IndexPackage<'_>>(line)?;
+            let summary = index.to_summary(source_id)?;
+            Ok((index, summary))
+        })();
+        let (index, summary, valid) = match index_summary {
+            Ok((index, summary)) => (index, summary, true),
+            Err(err) => {
+                let Ok(IndexPackageMinimum { name, vers }) =
+                    serde_json::from_slice::<IndexPackageMinimum>(line)
+                else {
+                    // If we can't recover, prefer the original error
+                    return Err(err);
+                };
+                tracing::info!(
+                    "recoverying from failed parse of registry package {name}@{vers}: {err}"
+                );
+                let IndexPackageRustVersion { rust_version } =
+                    serde_json::from_slice::<IndexPackageRustVersion>(line).unwrap_or_default();
+                let IndexPackageV { v } =
+                    serde_json::from_slice::<IndexPackageV>(line).unwrap_or_default();
+                let index = IndexPackage {
+                    name,
+                    vers,
+                    rust_version,
+                    v,
+                    deps: Default::default(),
+                    features: Default::default(),
+                    features2: Default::default(),
+                    cksum: Default::default(),
+                    yanked: Default::default(),
+                    links: Default::default(),
+                };
+                let summary = index.to_summary(source_id)?;
+                (index, summary, false)
+            }
+        };
         let v = index.v.unwrap_or(1);
         tracing::trace!("json parsed registry {}/{}", index.name, index.vers);
-        let summary = index.to_summary(source_id)?;
 
         let v_max = if bindeps {
             INDEX_V_MAX + 1
@@ -742,6 +798,8 @@ impl IndexSummary {
 
         if v_max < v {
             Ok(IndexSummary::Unsupported(summary, v))
+        } else if !valid {
+            Ok(IndexSummary::Invalid(summary))
         } else if index.yanked.unwrap_or(false) {
             Ok(IndexSummary::Yanked(summary))
         } else {

--- a/src/cargo/sources/registry/index/mod.rs
+++ b/src/cargo/sources/registry/index/mod.rs
@@ -169,12 +169,7 @@ impl IndexSummary {
 
     /// Extract the package id from any variant
     pub fn package_id(&self) -> PackageId {
-        match self {
-            IndexSummary::Candidate(sum)
-            | IndexSummary::Yanked(sum)
-            | IndexSummary::Offline(sum)
-            | IndexSummary::Unsupported(sum, _) => sum.package_id(),
-        }
+        self.as_summary().package_id()
     }
 
     /// Returns `true` if the index summary is [`Yanked`].

--- a/src/cargo/sources/registry/index/mod.rs
+++ b/src/cargo/sources/registry/index/mod.rs
@@ -259,8 +259,36 @@ pub struct IndexPackage<'a> {
     pub v: Option<u32>,
 }
 
+impl IndexPackage<'_> {
+    fn to_summary(&self, source_id: SourceId) -> CargoResult<Summary> {
+        // ****CAUTION**** Please be extremely careful with returning errors, see
+        // `IndexSummary::parse` for details
+        let pkgid = PackageId::new(self.name.into(), self.vers.clone(), source_id);
+        let deps = self
+            .deps
+            .iter()
+            .map(|dep| dep.clone().into_dep(source_id))
+            .collect::<CargoResult<Vec<_>>>()?;
+        let mut features = self.features.clone();
+        if let Some(features2) = &self.features2 {
+            for (name, values) in features2 {
+                features.entry(*name).or_default().extend(values);
+            }
+        }
+        let mut summary = Summary::new(
+            pkgid,
+            deps,
+            &features,
+            self.links,
+            self.rust_version.clone(),
+        )?;
+        summary.set_checksum(self.cksum.clone());
+        Ok(summary)
+    }
+}
+
 /// A dependency as encoded in the [`IndexPackage`] index JSON.
-#[derive(Deserialize, Serialize)]
+#[derive(Deserialize, Serialize, Clone)]
 pub struct RegistryDependency<'a> {
     /// Name of the dependency. If the dependency is renamed, the original
     /// would be stored in [`RegistryDependency::package`].
@@ -706,32 +734,10 @@ impl IndexSummary {
         // between different versions that understand the index differently.
         // Make sure to consider the INDEX_V_MAX and CURRENT_CACHE_VERSION
         // values carefully when making changes here.
-        let IndexPackage {
-            name,
-            vers,
-            cksum,
-            deps,
-            mut features,
-            features2,
-            yanked,
-            links,
-            rust_version,
-            v,
-        } = serde_json::from_slice(line)?;
-        let v = v.unwrap_or(1);
-        tracing::trace!("json parsed registry {}/{}", name, vers);
-        let pkgid = PackageId::new(name.into(), vers.clone(), source_id);
-        let deps = deps
-            .into_iter()
-            .map(|dep| dep.into_dep(source_id))
-            .collect::<CargoResult<Vec<_>>>()?;
-        if let Some(features2) = features2 {
-            for (name, values) in features2 {
-                features.entry(name).or_default().extend(values);
-            }
-        }
-        let mut summary = Summary::new(pkgid, deps, &features, links, rust_version)?;
-        summary.set_checksum(cksum);
+        let index: IndexPackage<'_> = serde_json::from_slice(line)?;
+        let v = index.v.unwrap_or(1);
+        tracing::trace!("json parsed registry {}/{}", index.name, index.vers);
+        let summary = index.to_summary(source_id)?;
 
         let v_max = if bindeps {
             INDEX_V_MAX + 1
@@ -741,7 +747,7 @@ impl IndexSummary {
 
         if v_max < v {
             Ok(IndexSummary::Unsupported(summary, v))
-        } else if yanked.unwrap_or(false) {
+        } else if index.yanked.unwrap_or(false) {
             Ok(IndexSummary::Yanked(summary))
         } else {
             Ok(IndexSummary::Candidate(summary))

--- a/src/cargo/sources/registry/mod.rs
+++ b/src/cargo/sources/registry/mod.rs
@@ -834,6 +834,9 @@ impl<'gctx> Source for RegistrySource<'gctx> {
                                 summary.version()
                             );
                         }
+                        IndexSummary::Invalid(summary) => {
+                            tracing::debug!("invalid ({} {})", summary.name(), summary.version());
+                        }
                         IndexSummary::Offline(summary) => {
                             tracing::debug!("offline ({} {})", summary.name(), summary.version());
                         }

--- a/tests/testsuite/registry.rs
+++ b/tests/testsuite/registry.rs
@@ -2916,7 +2916,9 @@ fn ignore_invalid_json_lines_git() {
 
 fn ignore_invalid_json_lines() {
     Package::new("foo", "0.1.0").publish();
-    Package::new("foo", "0.1.1").invalid_json(true).publish();
+    Package::new("foo", "0.1.1")
+        .invalid_index_line(true)
+        .publish();
     Package::new("foo", "0.2.0").publish();
 
     let p = project()

--- a/tests/testsuite/registry.rs
+++ b/tests/testsuite/registry.rs
@@ -3012,10 +3012,13 @@ fn invalid_json_lines_error() {
         .with_stderr_data(str![[r#"
 [UPDATING] `dummy-registry` index
 [ERROR] failed to select a version for the requirement `foo = "^0.1.1"`
-candidate versions found which didn't match: 0.2.0, 0.1.0
+  version 0.1.3 requires cargo 1.2345
+  version 0.1.4 requires a Cargo version that supports index version 1000000000
+  version 0.1.5's index entry is invalid
+  version 0.1.6 requires a Cargo version that supports index version 1000000000
+  version 0.1.7's index entry is invalid
 location searched: `dummy-registry` index (which is replacing registry `crates-io`)
 required by package `a v0.5.0 ([ROOT]/foo)`
-perhaps a crate was updated and forgotten to be re-vendored?
 
 "#]])
         .run();
@@ -3024,10 +3027,13 @@ perhaps a crate was updated and forgotten to be re-vendored?
         .with_stderr_data(str![[r#"
 [UPDATING] `dummy-registry` index
 [ERROR] failed to select a version for the requirement `foo = "^0.1.1"`
-candidate versions found which didn't match: 0.2.0, 0.1.0
+  version 0.1.3 requires cargo 1.2345
+  version 0.1.4 requires a Cargo version that supports index version 1000000000
+  version 0.1.5's index entry is invalid
+  version 0.1.6 requires a Cargo version that supports index version 1000000000
+  version 0.1.7's index entry is invalid
 location searched: `dummy-registry` index (which is replacing registry `crates-io`)
 required by package `a v0.5.0 ([ROOT]/foo)`
-perhaps a crate was updated and forgotten to be re-vendored?
 
 "#]])
         .run();

--- a/tests/testsuite/registry.rs
+++ b/tests/testsuite/registry.rs
@@ -2943,6 +2943,97 @@ fn ignore_invalid_json_lines() {
 }
 
 #[cargo_test]
+fn invalid_json_lines_error() {
+    Package::new("foo", "0.1.0")
+        .rust_version("1.0")
+        .schema_version(2)
+        .publish();
+    Package::new("foo", "0.1.1")
+        // Bad name field, too corrupt to use
+        .invalid_index_line(true)
+        .publish();
+    Package::new("foo", "0.1.2")
+        // Bad version field, too corrupt to use
+        .index_line(
+            r#"{"cksum":"7ca5fc2301ad96ade45356faf53225aea36437d99930bbfa951155c01faecf79","deps":[],"features":{},"links":null,"name":"foo","vers":"bad","yanked":false,"rust_version":"1.2345","v":1000000000}"#,
+        )
+        .publish();
+    Package::new("foo", "0.1.3")
+        // Bad field, report rust version
+        .index_line(
+            r#"{"cksum":"7ca5fc2301ad96ade45356faf53225aea36437d99930bbfa951155c01faecf79","deps":[],"features":"bad","links":null,"name":"foo","vers":"0.1.3","yanked":false,"rust_version":"1.2345","v":1000000000}"#,
+        )
+        .publish();
+    Package::new("foo", "0.1.4")
+        // Bad field, report schema
+        .index_line(
+            r#"{"cksum":"7ca5fc2301ad96ade45356faf53225aea36437d99930bbfa951155c01faecf79","deps":[],"features":"bad","links":null,"name":"foo","vers":"0.1.4","yanked":false,"v":1000000000}"#,
+        )
+        .publish();
+    Package::new("foo", "0.1.5")
+        // Bad field, report error
+        .index_line(
+            r#"{"cksum":"7ca5fc2301ad96ade45356faf53225aea36437d99930bbfa951155c01faecf79","deps":[],"features":"bad","links":null,"name":"foo","vers":"0.1.5","yanked":false}"#,
+        )
+        .publish();
+    Package::new("foo", "0.1.6")
+        // Bad field with bad rust version, report schema
+        .index_line(
+            r#"{"cksum":"7ca5fc2301ad96ade45356faf53225aea36437d99930bbfa951155c01faecf79","deps":[],"features":"bad","links":null,"name":"foo","vers":"0.1.6","yanked":false,"rust_version":"bad","v":1000000000}"#,
+        )
+        .publish();
+    Package::new("foo", "0.1.7")
+        // Bad field with bad rust version and schema, report error
+        .index_line(
+            r#"{"cksum":"7ca5fc2301ad96ade45356faf53225aea36437d99930bbfa951155c01faecf79","deps":[],"features":"bad","links":null,"name":"foo","vers":"0.1.7","yanked":false,"rust_version":"bad","v":"bad"}"#,
+        )
+        .publish();
+    Package::new("foo", "0.2.0").publish();
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "a"
+                version = "0.5.0"
+                edition = "2015"
+                authors = []
+
+                [dependencies]
+                foo = "0.1.1"
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    p.cargo("generate-lockfile")
+        .with_status(101)
+        .with_stderr_data(str![[r#"
+[UPDATING] `dummy-registry` index
+[ERROR] failed to select a version for the requirement `foo = "^0.1.1"`
+candidate versions found which didn't match: 0.2.0, 0.1.0
+location searched: `dummy-registry` index (which is replacing registry `crates-io`)
+required by package `a v0.5.0 ([ROOT]/foo)`
+perhaps a crate was updated and forgotten to be re-vendored?
+
+"#]])
+        .run();
+    p.cargo("generate-lockfile")
+        .with_status(101)
+        .with_stderr_data(str![[r#"
+[UPDATING] `dummy-registry` index
+[ERROR] failed to select a version for the requirement `foo = "^0.1.1"`
+candidate versions found which didn't match: 0.2.0, 0.1.0
+location searched: `dummy-registry` index (which is replacing registry `crates-io`)
+required by package `a v0.5.0 ([ROOT]/foo)`
+perhaps a crate was updated and forgotten to be re-vendored?
+
+"#]])
+        .run();
+}
+
+#[cargo_test]
 fn readonly_registry_still_works_http() {
     let _server = setup_http();
     readonly_registry_still_works();

--- a/tests/testsuite/registry.rs
+++ b/tests/testsuite/registry.rs
@@ -3235,6 +3235,17 @@ required by package `foo v0.1.0 ([ROOT]/foo)`
 
 "#]])
         .run();
+    p.cargo("generate-lockfile")
+        .with_status(101)
+        .with_stderr_data(str![[r#"
+[UPDATING] `dummy-registry` index
+[ERROR] failed to select a version for the requirement `bar = "^1.0"`
+  version 1.0.1 requires a Cargo version that supports index version 4294967295
+location searched: `dummy-registry` index (which is replacing registry `crates-io`)
+required by package `foo v0.1.0 ([ROOT]/foo)`
+
+"#]])
+        .run();
 }
 
 #[cargo_test]

